### DISCUSSION TITLE
How to get error line number while executing ruby code.

### DISCUSSION THREAD

**Discussion by Hareramrai** [opal/opal#2464] [Q&A]
I am opal in my application but I am having a problem with getting an error line number while executing any code.

````ruby
source = <<~EOS
puts "Hi"
puts "Hello" + 2
puts "Welcome"
EOS

code = Opal.compile(source)
result = `eval(js_code)`
````
In a real scenario, we are getting this source code from a web-based code editor.

Please help to get the error line number.

We are building something similar to TryRuby.

**Comment by hmdne** [comment]
I don't think that's an easy problem to be solved, but hey, we may at least try.

There's no mapping of the lines in the generated code to the lines of the source code. You can generate it alongside, it's called a source map.

```ruby
require 'opal/source_map'
c = Opal::Compiler.new("puts 'hello world'\nputs 'error world'+3\nputs 'unreachable'")
c.compile # Generates the code
c.source_map # Generates the source map
```

The thing that we get out of that is an object of a class Opal::SourceMap::File.

https://github.com/opal/opal/blob/master/lib/opal/source_map/file.rb

Now, it's easy to generate a source map in a format understandable by browsers. Let's try to eval it this way:

```ruby
`eval(#{c.compile+"\n"+c.source_map.to_data_uri_comment})`
```

This could give us some hint to what's happening, but it depends on an implementation to support that. You may get some info in a backtrace. But you may not. I get this backtrace on opal-repl (miniracer):

```
	JavaScript at Function.$$type_error (<anonymous>:2445:42)
	JavaScript at Function.$$coerce_to (<anonymous>:2456:26)
	JavaScript at String.$String_$plus$7 (<anonymous>:6486:36)
	JavaScript at $rb_plus (eval at <anonymous> (unknown source), <anonymous>:4:90)
	JavaScript at eval (eval at <anonymous> (unknown source), <anonymous>:11:14)
	JavaScript at eval (eval at <anonymous> (unknown source), <anonymous>:13:3)
(...)
```

The thing we are interested in is obviously:

```
	JavaScript at eval (eval at <anonymous> (unknown source), <anonymous>:11:14)
```

(Obviously for us, for your implementation it may not be that obvious).

That line is:
```
>> c.compile.split("\n")[11-1]
=> "  self.$puts($rb_plus(\"error world\", 3));"
>> 
```

Ok, let's try some more on this source_map object. It has some methods that look promising enough to suspect that they contain the source map in a more unprocessed form:

```
>> c.source_map.absolute_mappings[11-1]
=> [[0, 0, 0, 0], [6, 0, 1, 0, 0], [12, 0, 1, 0, 0], [13, 0, 1, 5, 1], [22, 0, 1, 5], [35, 0, 1, 5, 1], [37, 0, 1, 19, 2], [38, 0, 1, 5, 1], [39, 0, 1, 0, 0], [40, 0, 0, 0]]
>> 
```

Ok, we are getting somewhere. Now, I don't know anything about this format, but let's try some more. We got a line 11 and character 14. Computers start counting characters from 0, so it's character 13. And this thing has 13:

```
[13, 0, >>1<<, 5, 1]
```

Was the original code on line 2 (1+1)? Yes. So, let's try that on a different example:

```
>> src = "puts 'hello world'\n\n\n\nputs 'xx'\n\n\n\nputs 'error world hello'+3\nputs 'unreachable'"
=> "puts 'hello world'\n\n\n\nputs 'xx'\n\n\n\nputs 'error world hello'+3\nputs 'unreachable'"
>> c = Opal::Compiler.new(src)
=> #<Opal::Compiler:0x394c>
>> `eval(#{c.compile})`
TypeError: no implicit conversion of Number into String
	JavaScript at Function.$$type_error (<anonymous>:2445:42)
	JavaScript at Function.$$coerce_to (<anonymous>:2456:26)
	JavaScript at String.$String_$plus$7 (<anonymous>:6486:36)
	JavaScript at $rb_plus (eval at <anonymous> (unknown source), <anonymous>:4:90)
	JavaScript at eval (eval at <anonymous> (unknown source), <anonymous>:12:14)
	JavaScript at eval (eval at <anonymous> (unknown source), <anonymous>:14:3)
        (...)
=> 
>> c.source_map.absolute_mappings[12-1]
=> [[0, 0, 0, 0], [6, 0, 8, 0, 0], [12, 0, 8, 0, 0], [13, 0, 8, 5, 1], [22, 0, 8, 5], [41, 0, 8, 5, 1], [43, 0, 8, 25, 2], [44, 0, 8, 5, 1], [45, 0, 8, 0, 0], [46, 0, 0, 0]]
>> # We are interested in [13, 0, >>8<<, 5, 1], because character in a backtrace is 14-1
>> src.split("\n")[8]
=> "puts 'error world hello'+3"
>> 
```

It worked? Well it did :D. I know it may not be a solution you were looking for and maybe you think that this should be a part of the corelib and readable from an Exception object. We did as bit of hackery here and including this thing by default would increase the generated source codes (we would need to include the absolute_mappings along with the JS compiled code), which isn't what most Opal users want, but you can surely use it in your implementation.

**Comment by Hareramrai** [comment]
Thanks, @hmdne for a detailed response. 

I tried the same approach for the below code snippet but not able to get the exact line number.

````ruby


# Calculate grade using if and range

puts "John, how many questions you got right?"

answer = 30
if answer > 100
  puts "That’s not possible"
end
if answer == 100
  puts "You got A+ grade."
end
if (80..100).include? answer
  puts "You got A grade"
end
if (60..81).include? answer
  puts "You got B grade"
end
if (40..61).include? answer
  puts "You got C grade"
end
if (20..41).include? answer
  puts "You got D grade" + 4
end
if (0..21).include? answer
  puts "You got E grade"
end
if answer == 0
  puts "You got F grade"
end

````

I got the below error. 
````
"TypeError: no implicit conversion of Number into String
    at Function.$type_error (http://localhost:5000/assets/corelib/helpers.self-91af85577d91c2ba2d161f300d4440f43572f840c8475c6dfbaf1ae7f50ae9a1.js?body=1:32:42)
    at Function.$coerce_to (http://localhost:5000/assets/corelib/helpers.self-91af85577d91c2ba2d161f300d4440f43572f840c8475c6dfbaf1ae7f50ae9a1.js?body=1:43:26)
    at String.$+ (http://localhost:5000/assets/corelib/string.self-e27a8e8a7942006ed2f44ef9c3b74d688634c8557776fbf918f1b60c4c5ac8da.js?body=1:125:36)
    at $rb_plus (eval at $$execute_ruby_code (http://localhost:5000/assets/opal_configuration.self-61a0a90bb8ef3bbae7286cecde2bd68750201af903b9f3d5ec733f303cb6ce34.js?body=1:127:18), <anonymous>:7:90)
    at eval (eval at $$execute_ruby_code (http://localhost:5000/assets/opal_configuration.self-61a0a90bb8ef3bbae7286cecde2bd68750201af903b9f3d5ec733f303cb6ce34.js?body=1:127:18), <anonymous>:26:16)
    at eval (eval at $$execute_ruby_code (http://localhost:5000/assets/opal_configuration.self-61a0a90bb8ef3bbae7286cecde2bd68750201af903b9f3d5ec733f303cb6ce34.js?body=1:127:18), <anonymous>:34:3)
    at Function.$execute_ruby_code (http://localhost:5000/assets/opal_configuration.self-61a0a90bb8ef3bbae7286cecde2bd68750201af903b9f3d5ec733f303cb6ce34.js?body=1:127:18)
    at _this.handleRun (http://localhost:5000/packs/js/application-5013cf2e58057a3cb2bf.js:4529:28)
    at HTMLUnknownElement.callCallback (http://localhost:5000/packs/js/application-5013cf2e58057a3cb2bf.js:101012:18)
    at Object.invokeGuardedCallbackDev (http://localhost:5000/packs/js/application-5013cf2e58057a3cb2bf.js:101061:20)"
````

From the above error, I got 26 and 34 line numbers. 

Also, absolute mapping returns below array for 26-1. 
````
 puts compiler.source_map.absolute_mappings[25]
[[0, 0, 17, 0], [8, 0, 18, 2, 0], [14, 0, 18, 2, 0], [15, 0, 18, 7, 14], [24, 0, 18, 7], [41, 0, 18, 7, 14], [43, 0, 18, 27, 15], [44, 0, 18, 7, 14], [45, 0, 18, 2, 0], [46, 0, 17, 0], [47, 0, 0, 0]]
````
So for getting the exact error line number, I am not able to find an array that contains 34-1. 

Could you please help me with this.

**Comment by hmdne** [comment]
Of course. You need to take a line and character of a single backtrace entry into consideration.

It's `<anonymous>:26:16` and `<anonymous>:34:3`. The second one is uninteresting (it's about wrapping the generated code in a function call).

Therefore, it's this array that you are interested in, because it starts with 16-1: `[15, 0, 18, 7, 14]`. From this you can get "18" as the line number from Ruby source and possibly "7" means the character.

**Comment by elia** [comment]
depending on the environment in which the code is run you can also try with https://github.com/evanw/node-source-map-support or other project that will sort out the mapping and backtrace for you… 

Another option is to inline the generated source at the end of the eval and let the browser do the work, using builder the code looks like this: `builder.to_s + "\n" + builder.source_map.to_data_uri_comment`, the same can be done with the compiler.

This inlined sourcemap will also contain the full-body of the original ruby sources so it should be complete.
